Panic with unread show series iterators during drop database

### Bug report

__System info:__ 

OSX, built from 7e515cf5f7c298abd8aae62ed1f7bb85601a4b4d

__Steps to reproduce:__

Run repro script here:
https://gist.github.com/desa/1d43052f99b44f8957b3194d69c0588f

__Expected behavior:__ 
Database is dropped correctly.

__Actual behavior:__ 

This panic:

```
[query] 2016/09/20 14:12:16 SELECT "key" FROM stress.autogen._series
[query] 2016/09/20 14:12:16 DROP DATABASE stress
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x24a699]

goroutine 491 [running]:
panic(0x7b07c0, 0xc82000a190)
       	/usr/local/Cellar/go/1.6.2/libexec/src/runtime/panic.go:481 +0x3e6
github.com/influxdata/influxdb/tsdb.(*Measurement).AppendSeriesKeysByID(0xc8369db8c0, 0xc8375a6c30, 0x0, 0x1, 0xc8369fad70, 0x1, 0x1, 0x0, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/meta.go:600 +0x239
github.com/influxdata/influxdb/tsdb.(*seriesIterator).nextKeys(0xc837bfc780, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/shard.go:1050 +0x1ed
github.com/influxdata/influxdb/tsdb.(*seriesIterator).Next(0xc837bfc780, 0xc83cc83290, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/shard.go:1007 +0x66
github.com/influxdata/influxdb/influxql.(*floatSortedMergeIterator).pop(0xc837a3a240, 0xd9f670c6d84be601, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:388 +0x3fd
github.com/influxdata/influxdb/influxql.(*floatSortedMergeIterator).Next(0xc837a3a240, 0x13b28, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:350 +0x2d
github.com/influxdata/influxdb/influxql.(*floatInterruptIterator).Next(0xc83edacf00, 0x663400, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:761 +0xb5
github.com/influxdata/influxdb/influxql.(*floatFastDedupeIterator).Next(0xc83edacf20, 0x1, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/influxql/iterator.go:1191 +0x5a
github.com/influxdata/influxdb/influxql.(*bufFloatIterator).Next(0xc83edacf40, 0x1, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:96 +0x69
github.com/influxdata/influxdb/influxql.(*floatAuxIterator).stream(0xc83c8d6000)
       	/Users/mr/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:880 +0x2b
created by github.com/influxdata/influxdb/influxql.(*floatAuxIterator).Start
       	/Users/mr/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:844 +0x35
exit status 2
```

__Additional info:__ 

The repro script opens an HTTP request for `SHOW SERIES`, but does not read or close the response body. Then a separate HTTP request runs `DROP DATABASE`. Eventually the iterator for show series advances, but it tries to read a shard that was removed when the database was dropped, and it panics there.

cc @jsternberg

It still panics, although in my experience today, it either fails on the first couple runs from a fresh start of influxdb, or it succeeds forever. 

I tried to write a shorter repro with only the standard library, but I had less luck reproducing the issue with that approach. Here's the stack trace on master at b385decc06c549fb8550b901b53c24e4b4fe46bf. I had to use the repro script in the gist, with influx-stress at commit 96012d6c5d72fdcc61bdafc029214d0b786e5e0b. 

```
[query] 2016/10/18 09:03:47 DROP DATABASE stress
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x246e89]

goroutine 755 [running]:
panic(0x7c04a0, 0xc82000a190)
       	/usr/local/Cellar/go/1.6.2/libexec/src/runtime/panic.go:481 +0x3e6
github.com/influxdata/influxdb/tsdb.(*Measurement).AppendSeriesKeysByID(0xc8215791a0, 0xc82331a410, 0x0, 0x1, 0xc8216fc668, 0x1, 0x1, 0x0, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/meta.go:601 +0x239
github.com/influxdata/influxdb/tsdb.(*seriesIterator).nextKeys(0xc821d98f00, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/shard.go:1203 +0x1ed
github.com/influxdata/influxdb/tsdb.(*seriesIterator).Next(0xc821d98f00, 0xc82f21c650, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/shard.go:1160 +0x66
github.com/influxdata/influxdb/influxql.(*floatSortedMergeIterator).pop(0xc82ef75680, 0xc82f1f6001, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:388 +0x3fd
github.com/influxdata/influxdb/influxql.(*floatSortedMergeIterator).Next(0xc82ef75680, 0x134b8, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:350 +0x2d
github.com/influxdata/influxdb/influxql.(*floatInterruptIterator).Next(0xc82dd68d80, 0x671120, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:777 +0xb5
github.com/influxdata/influxdb/influxql.(*floatFastDedupeIterator).Next(0xc82dd68dc0, 0x1, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/influxql/iterator.go:1191 +0x5a
github.com/influxdata/influxdb/influxql.(*bufFloatIterator).Next(0xc82dd68e00, 0x1, 0x0, 0x0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:96 +0x69
github.com/influxdata/influxdb/influxql.(*floatAuxIterator).stream(0xc82f10d9e0)
       	/Users/mr/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:896 +0x2b
created by github.com/influxdata/influxdb/influxql.(*floatAuxIterator).Start
       	/Users/mr/go/src/github.com/influxdata/influxdb/influxql/iterator.gen.go:860 +0x35
exit status 2
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -48,6 +48,7 @@
 - [#7385](https://github.com/influxdata/influxdb/pull/7385): Reduce query planning allocations
 - [#7436](https://github.com/influxdata/influxdb/issues/7436): Remove accidentally added string support for the stddev call.
 - [#7161](https://github.com/influxdata/influxdb/issues/7161): Drop measurement causes cache max memory exceeded error.
+- [#7334](https://github.com/influxdata/influxdb/issues/7334): Panic with unread show series iterators during drop database
 
 ## v1.0.2 [2016-10-05]
 

--- a/tsdb/meta.go
+++ b/tsdb/meta.go
@@ -598,7 +598,9 @@ func (m *Measurement) AppendSeriesKeysByID(dst []string, ids []uint64) []string 
 	m.mu.RLock()
 	defer m.mu.RUnlock()
 	for _, id := range ids {
-		dst = append(dst, m.seriesByID[id].Key)
+		if s := m.seriesByID[id]; s != nil {
+			dst = append(dst, s.Key)
+		}
 	}
 	return dst
 }

--- a/tsdb/meta_test.go
+++ b/tsdb/meta_test.go
@@ -12,7 +12,7 @@ import (
 )
 
 // Test comparing SeriesIDs for equality.
-func Test_SeriesIDs_Equals(t *testing.T) {
+func TestSeriesIDs_Equals(t *testing.T) {
 	ids1 := tsdb.SeriesIDs([]uint64{1, 2, 3})
 	ids2 := tsdb.SeriesIDs([]uint64{1, 2, 3})
 	ids3 := tsdb.SeriesIDs([]uint64{4, 5, 6})
@@ -25,7 +25,7 @@ func Test_SeriesIDs_Equals(t *testing.T) {
 }
 
 // Test intersecting sets of SeriesIDs.
-func Test_SeriesIDs_Intersect(t *testing.T) {
+func TestSeriesIDs_Intersect(t *testing.T) {
 	// Test swaping l & r, all branches of if-else, and exit loop when 'j < len(r)'
 	ids1 := tsdb.SeriesIDs([]uint64{1, 3, 4, 5, 6})
 	ids2 := tsdb.SeriesIDs([]uint64{1, 2, 3, 7})
@@ -48,7 +48,7 @@ func Test_SeriesIDs_Intersect(t *testing.T) {
 }
 
 // Test union sets of SeriesIDs.
-func Test_SeriesIDs_Union(t *testing.T) {
+func TestSeriesIDs_Union(t *testing.T) {
 	// Test all branches of if-else, exit loop because of 'j < len(r)', and append remainder from left.
 	ids1 := tsdb.SeriesIDs([]uint64{1, 2, 3, 7})
 	ids2 := tsdb.SeriesIDs([]uint64{1, 3, 4, 5, 6})
@@ -71,7 +71,7 @@ func Test_SeriesIDs_Union(t *testing.T) {
 }
 
 // Test removing one set of SeriesIDs from another.
-func Test_SeriesIDs_Reject(t *testing.T) {
+func TestSeriesIDs_Reject(t *testing.T) {
 	// Test all branches of if-else, exit loop because of 'j < len(r)', and append remainder from left.
 	ids1 := tsdb.SeriesIDs([]uint64{1, 2, 3, 7})
 	ids2 := tsdb.SeriesIDs([]uint64{1, 3, 4, 5, 6})
@@ -90,6 +90,32 @@ func Test_SeriesIDs_Reject(t *testing.T) {
 
 	if !exp.Equals(got) {
 		t.Fatalf("exp=%v, got=%v", exp, got)
+	}
+}
+
+func TestMeasurement_AppendSeriesKeysByID_Missing(t *testing.T) {
+	m := tsdb.NewMeasurement("cpu")
+	var dst []string
+	dst = m.AppendSeriesKeysByID(dst, []uint64{1})
+	if exp, got := 0, len(dst); exp != got {
+		t.Fatalf("series len mismatch: exp %v, got %v", exp, got)
+	}
+}
+
+func TestMeasurement_AppendSeriesKeysByID_Exists(t *testing.T) {
+	m := tsdb.NewMeasurement("cpu")
+	s := tsdb.NewSeries("cpu,host=foo", models.Tags{models.Tag{Key: []byte("host"), Value: []byte("foo")}})
+	s.ID = 1
+	m.AddSeries(s)
+
+	var dst []string
+	dst = m.AppendSeriesKeysByID(dst, []uint64{1})
+	if exp, got := 1, len(dst); exp != got {
+		t.Fatalf("series len mismatch: exp %v, got %v", exp, got)
+	}
+
+	if exp, got := "cpu,host=foo", dst[0]; exp != got {
+		t.Fatalf("series mismatch: exp %v, got %v", exp, got)
 	}
 }
 

